tools/maplint/source/dmm.py

```diff
@@ -0,0 +1,178 @@
+# I know we already have one in mapmerge, but this one can afford to be significantly simpler to interface with
+# by virtue of being read-only.
+import re
+from dataclasses import dataclass, field
+
+from .common import Constant, Filename, Null, Typepath
+from .error import MaplintError
+
+REGEX_POP_ID = re.compile(r'^"(?P<key>.+)" = \($')
+REGEX_POP_CONTENT_HEADER = re.compile(r'^(?P<path>[/\w]+?)(?P<end>[{,)])$')
+REGEX_ROW_BEGIN = re.compile(r'^\((?P<x>\d+),(?P<y>\d+),(?P<z>\d+)\) = {"$')
+REGEX_VAR_EDIT = re.compile(r'^\t(?P<name>.+?) = (?P<definition>.+?);?$')
+
+@dataclass
+class Content:
+    path: Typepath
+    var_edits: dict[str, Constant] = field(default_factory = dict)
+
+@dataclass
+class DMM:
+    pops: dict[str, list[Content]] = field(default_factory = dict)
+
+    # Z -> X -> Y -> Pop
+    turfs: list[list[list[str]]] = field(default_factory = list)
+
+    def size(self):
+        return (len(self.turfs[0]), len(self.turfs[0][0]))
+
+    def turfs_for_pop(self, key: str):
+        for z, z_level in enumerate(self.turfs):
+            for x, x_level in enumerate(z_level):
+                for y, turf in enumerate(x_level):
+                    if turf == key:
+                        yield (x, y, z)
+
+class DMMParser:
+    dmm: DMM
+    line = 0
+
+    def __init__(self, reader):
+        self.dmm = DMM()
+        self.reader = reader
+
+    def parse(self):
+        if "dmm2tgm" not in self.next_line():
+            raise MaplintError("Map isn't in TGM format. Consider using StrongDMM instead of Dream Maker.")
```

Comment:
```suggestion
            raise MaplintError("Map isn't in TGM format. Consider using StrongDMM instead of Dream Maker. Please also look into installing the map merge tools.")
```

The way I've seen it for a good while: StrongDMM isn't a fix-all for this issue- while it does default to the TGM format setting when you start editing, there is an option to select the DMM format in the settings for StrongDMM (for whatever reason), since people might edit it un-necessarily or have it edited to be complaint somewhere else (silly, but could happen). If they have the map merge tools installed for their local repository, they will always push commits in the TGM format regardless of local settings, which is much more preferable IMO. We also have control over the TGM format with the map-merge tools, so should something need to change for whatever reason, it's best that everyone's already got the git hooks installed so they can immediately reap the benefits.

Don't know if linking to https://github.com/tgstation/tgstation/tree/master/tools/hooks would be appropriate in such a message.

Reply:
The git hooks break constantly for me because they rely on the extremely fallible bootstrapping script, which I had two non-trivial to fix problems with just while trying to open this PR. I would definitely disagree that they are the correct fix-all for this.

I think I pinged you in on this (but not Vire) about how I'm going to be investigating a tg specific fork of StrongDMM in the future, so we could probably remove the option of this then.

```suggestion
            raise MaplintError("Map isn't in TGM format. Consider using StrongDMM instead of Dream Maker.\n  Please also consider installing the map merge tools, found through Install.bat in the tools/hooks folder.")
```

Reply:
Mapmerge isn't perfect, but it does catch a large number of issues. Not perfectly though. I've seen at least one circumstance where a map has been fully converted from tgm to dmm, committed, then reconverted and committed again. Huge mess.

I'm curious what would be added to a tg fork of StrongDMM though? Are you interested in enforcing these at the editor level? Or just providing more interactive feedback for the mapper?

Reply:
I sent you a DM on Discord before you sent the message, it's something that I plan on fleshing out more in the future but have just been randomly musing about in other Discords @Maurukas